Convert "Reasoning with inductive types" chapter to prodn. (#14179)
coq/coq: doc/sphinx/proofs/writing-proofs/inductive.rst

Applying constructors
---------------------

.. tacn:: constructor {? @nat_or_var } {? with @bindings }

   .. todo PR: This really does an intros, peeling off foralls and -> ?
      If @nat_or_var is not given, it just picks the right contructor?

>       (in which case, why would you ever bother to give @nat_or_var?).

Reviewer: >This really does an intros, peeling off foralls and -> ?

Yes.

>If `@nat_or_var` is not given, it just picks the right contructor?

It picks one constructor that can be applied to the current goal, but it's not always the "right" one. So in some cases, you should be explicit on the constructor you want to apply.

E.g.:

```coq
Goal forall A B : Prop, B -> A \/ B.
constructor. (* does the same thing as constructor 1 / left, because it succeeds,
                but it makes the goal unprovable *)
(*
1 subgoal
A, B : Prop
H : B
______________________________________(1/1)
A
*)
Undo.
constructor 2. (* does the same thing as right *)
(*
1 subgoal
A, B : Prop
H : B
______________________________________(1/1)
B
*)

Author: Uh.  Backtracking may complicate the explanation of many, many tactics.  Any feel for how many tactics can produce multiple results with backtracking?  Maybe this topic deserves a bit more explanation somewhere (and not just in the Ltac chapter).

Here's what I have now:

If not specified, the tactic tries all the constructors,
   which can result in more than one match (e.g. for `\\/`).

Reviewer 2: >  Any feel for how many tactics can produce multiple results with backtracking?

constructor, multimatch, +, typeclasses eauto.  `;` will preserve multiple successes but not introduce new ones.  I think this might be it?

Reviewer: I think the solution is to write that `constructor` may produce more than one success with "success" a glossary term (linking back to the Ltac chapter for the readers that want to understand what "producing more than one success" means).

The tactics presented here specialize :tacn:`apply` (and
:tacn:`eapply`) to the case of constructors of inductive types.

.. tacn:: constructor {? @nat_or_var } {? with @bindings }

   .. todo PR: This really does an intros, peeling off foralls and -> ?
      If @nat_or_var is not given, it just picks the right contructor?
      (in which case, why would you ever bother to give @nat_or_var?).

>    When the goal conclusion is an inductive type `I`, this is equivalent

Author: Can you give me precise language for your informal description?

>    When the goal conclusion is an inductive type `I`, this is equivalent

Reviewer 2: I think there's some language describing the `red` tactic that can be reused.  Here is some rough but precise language describing the tactic: Repeat the following process: if the whnf of the goal is an inductive type, apply the given constructor; if it is a forall, run `intro`; otherwise, fail.  Said another way, if after running `repeat intro; hnf`, the goal is an inductive type then `apply` the ith constructor, otherwise fail.

>    .. tacv:: constructor

Author: I haven't finished removing the tactic variants, nor have I pushed any of the removals.

      .. exn:: Not an inductive goal with 1 constructor.
         :undocumented:

   .. tacn:: left {? with @bindings }
             right {? with @bindings }

      These tactics apply only if :g:`I` has two constructors, for
      instance in the case of a disjunction `A \/ B`.
      Then they are respectively equivalent to
      :n:`constructor 1 {? with @bindings }` and
      :n:`constructor 2 {? with @bindings }`.

      .. exn:: Not an inductive goal with 2 constructors.
         :undocumented:

   .. tacn:: econstructor {? @nat_or_var {? with @bindings } }

>              eexists {? {+, @bindings } }

Author: I think the syntax is correct.  "exists" doesn't require an argument, e.g.

```
Goal True.
exists.  (* No more subgoals *)
```

Superficially, it looks a little odd that all of these have `with @bindings` except for `(e)exists`.

>              eexists {? {+, @bindings } }

Reviewer 2: I think I was specifically commenting on that we previously had `exists {*, @bindings }` but `eexists {? {+, @bindings } }` and it seemed to me that `{*, @bindings }` and `{? {+, @bindings } }` ought to express the same thing but seem superficially different.

Also, I was not previously aware that `exists` does not require an argument; neat!

> Superficially, it looks a little odd that all of these have `with @bindings` except for `(e)exists`.

I agree that it's weird superficially, but I think that it fits with the deeper meaning.  `with` is for optional arguments which are not the main point of the tactic, whereas for `(e)exists`, the bindings are the main focus.

The tactics presented in this section implement induction or case
analysis on inductive or co-inductive objects (see :ref:`inductive-definitions`).

.. tacn:: destruct @induction_clause_list

   .. insertprodn induction_clause_list opt_clause

   .. prodn::
      induction_clause_list ::= {+, @induction_clause } {? using @one_term {? with @bindings } } {? @opt_clause }
      induction_clause ::= @destruction_arg {? @as_or_and_ipat } {? eqn : @naming_intropattern } {? @opt_clause }
      opt_clause ::= in @goal_occurrences
      | at @occs_nums

   .. todo PR: need to explain how to get @term from @induction_clause_list
      and @destruction_arg for several tactics,

>       also what it means if multiple @induction_clauses are given

Reviewer: @JasonGross So I guess you actually use it in practice and would not like to see this form deprecated? Because that's something that I've wanted to do for some time. (I'm talking about `induction x, y` here, not `destruct x, y`.)

BTW do you understand what `induction x, y using foo` does? I was expecting that the induction principle passed with the `using` clause would only be used for the first variable (consistently with the behavior when the `using` clause is not there) but if that was the case, then it wouldn't fail here:

```coq
Goal forall n m, n + m = m + n.
induction n, m using nat_ind. (* Error: Don't know where to find some argument. *)
```

>       also what it means if multiple @induction_clauses are given

Reviewer 2: > So I guess you actually use it in practice

Yes

>  and would not like to see this form deprecated?

Eh, I won't object strongly, `induction x; destruct y` is not that hard to type.  But I do have a slight preference for keeping it.

> BTW do you understand what `induction x, y using foo` does? I was expecting that the induction principle passed with the `using` clause would only be used for the first variable (consistently with the behavior when the `using` clause is not there) but if that was the case, then it wouldn't fail here:
> 
> ```coq
> Goal forall n m, n + m = m + n.
> induction n, m using nat_ind. (* Error: Don't know where to find some argument. *)
> ```

Heh, I don't.  Even more strange is that you get a different and nonsensically incomplete error message if you run intros first:
```coq
Goal forall n m, n + m = m + n.
  intros.
  destruct n, m using nat_ind. (* Error: Not the right number of induction arguments (expected ). *)
```
Interestingly, I can get the same message with
```coq
Goal forall n m, n + m = m + n.
  intros.
  destruct n using (nat_case _). (* Error: Not the right number of induction arguments (expected ). *)
```

>       also what it means if multiple @induction_clauses are given

Reviewer 2: Reported this issue as https://github.com/coq/coq/issues/14454.  I suspect it may be related to the issue you see without `intros`.  Perhaps @herbelin knows?

>       also what it means if multiple @induction_clauses are given

Reviewer 3: > Reported this issue as #14454. I suspect it may be related to the issue you see without `intros`. Perhaps @herbelin knows?

Here is an example of application:
```
Require Import Arith.
Goal forall a b, a +b =0.
intros.
induction a, b using lt_wf_double_rect.
(*
  b1, b2 : nat
  H : forall p q : nat, p < b1 -> p + q = 0
  H0 : forall p : nat, p < b2 -> b1 + p = 0
  ============================
  b1 + b2 = 0
*)
```
When using `using`, as many arguments as the number of arguments of the scheme is expected (plus a special rule to simulate `functional induction` which uses `using` under the hood).

>       also what it means if multiple @induction_clauses are given

Author: Is this also a bug (for `induction a, b.`?)

```
Goal forall a b, a + b = 0.
induction a, b using nat_ind.  (* Don't know where to find some argument. *)

induction a, b.
(*
4 subgoals
______________________________________(1/4)
0 + 0 = 0
______________________________________(2/4)
0 + S b = 0
______________________________________(3/4)
S a + 0 = 0
______________________________________(4/4)
S a + S b = 0
*)

(* FWIW, not equivalent to this: *)
induction a. induction b.
(* 3 subgoals
______________________________________(1/3)
0 + 0 = 0
______________________________________(2/3)
0 + S b = 0
______________________________________(3/3)
forall b : nat, S a + b = 0 *)
```

>       also what it means if multiple @induction_clauses are given

Reviewer: As discussed, `induction x, y` and `induction x, y using ...` really have nothing to do with each other. While the first is a shortcut for `induction x; destruct y`, the second expects an induction principle which talks about predicates of two variables (like in Hugo's example). I'm not a fan of this discrepancy, but since it exists (and is being relied on), let's document it.

      is erased (to avoid erasure, use parentheses, as in :n:`destruct (@ident)`).

      If :token:`ident` is a hypothesis of the context, and :token:`ident`
      is no longer dependent in the goal after application

>       of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as

Author: What is the difference in treatment?  Is the difference specific to `destruct` or is it more general?  Is this point mentioned elsewhere in the doc?

>       of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as

Reviewer 2: https://github.com/coq/coq/issues/2901  I'm not sure if there are other tactics that with the same bug

>       of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as

Reviewer: Section variables are still giving headaches to developers every time the topic comes back on the table.

      in :n:`destruct (@ident)`).

   .. tacv:: destruct @natural

>       :n:`destruct @natural` behaves like :n:`intros until @natural`

Author: Then it should be "behaves [exactly] the same as" or "is equivalent to".  "behaves as" is not grammatical.  It doesn't mean "behaves exactly the same as" and seems rather strange to me.  Agreed?

https://www.grammarly.com/blog/like-vs-as/ (The doc is formal writing, not a personal communication).

https://www.google.com/search?q=as+definition&rlz=1C1CHBF_enUS860US860&oq=as+definition&aqs=chrome..69i57j0l6j69i60.3735j1j7&sourceid=chrome&ie=UTF-8  (Oxford)

https://www.dictionary.com/browse/as (Dictionary.com)

   .. tacv:: destruct {+, @term}

      This is a shortcut for :n:`destruct @term; …; destruct @term`.

   .. tacv:: destruct @term @as_or_and_ipat

>       Behaves like :n:`destruct @term` but uses the names

Author: > it can be `destruct x as [ -> ? <- ]`

Readers can get this by following the grammar for `or_and_intropattern` though it is a couple levels down :-(.

      This behaves like :n:`destruct @term` providing explicit instances for

>       the dependent premises of the type of :token:`term`.

Reviewer: "free variables" do not really apply here. These variables are quantified with a `forall`. This is why the term "dependent premises" is used here. I'm not sure whether there's a better alternative. Maybe the solution is simply to define this term...

      clause is an occurrence clause whose syntax and behavior is described
      in :ref:`occurrences sets <occurrencessets>`.

   .. tacv:: destruct @term {? with @bindings } {? @as_or_and_ipat } {? eqn:@naming_intropattern } {? using @term {? with @bindings } } {? in @goal_occurrences }
             edestruct @term {? with @bindings } {? @as_or_and_ipat } {? eqn:@naming_intropattern } {? using @term {? with @bindings } } {? in @goal_occurrences }

      These are the general forms of :tacn:`destruct` and :tacn:`edestruct`.
      They combine the effects of the ``with``, ``as``, ``eqn:``, ``using``,
      and ``in`` clauses.

.. tacn:: case @induction_clause_list

   The tactic :n:`case` is a more basic tactic to perform case analysis without

>    recursion. It behaves like :n:`elim @term` but uses a case-analysis

Author: > @Zimmi48 Deprecate tactics that are not used anymore: `cut`, `elim`, `case` (vanilla versions), as well as `simple destruct` [from [#8946](https://github.com/coq/coq/issues/8946#issue-378672977)]

Is this a good time to deprecate `simple destruct`, which has only 9 uses in the standard library?  OTOH, I see ~3000 of uses of `elim` and ~800 of `cut`--do you really want to deprecate these (or recommend using other tactics in their place)?

If replacing of any of 4 mentioned tactics is mostly mechanical, I might create PRs to remove them for 8.15 though I might well need help for exceptional uses.

>    recursion. It behaves like :n:`elim @term` but uses a case-analysis

Reviewer: Why not raising the question of deprecating `simple destruct`. `cut` can really be replaced by `assert` or `enough` so it could make sense to deprecate it in theory, but it would be more work (I'm not sure how easy it would be to do this mechanically). It doesn't seem like `elim` can be deprecated at this time, but we can document that its use is discouraged for basic users.

>    elimination principle rather than a recursive one.

Author: Re:
```
.. tacn:: case @induction_clause_list   
The tactic :n:`case` is a more basic tactic to perform case analysis without
recursion. It behaves as :n:`elim @term` but uses a case-analysis
elimination principle and not a recursive one.
```

I'd like to understand this better because `simple destruct` is defined in terms of `case`.  Perhaps you can point me to a clear example?

Reviewer: The table to have in mind is:

|    |      Old (no `intros`, no `as`)      |  New (does `intros`, `as` clause) |
|----------|:-------------:|:-----:|
| no induction hypothesis (inserts a `match`) |  case | destruct |
| induction hypothesis (uses induction principle) |    elim   |   induction |

Beyond this, there are subtleties of implementation that I'm not knowledgeable about.

Reviewer 2: > Old (no `intros`, no `as`)

Also does not change uses of the variable in the context.

Author: The table is great as a mental model, though I don't feel it's necessary to include it as a table.  The syntax differences will be apparent and in the text I do label the old tactics as such, explain which use induction hypotheses, and which modify the local context.

.. tacv:: case @term with @bindings

   Analogous to :n:`elim @term with @bindings` above.

.. tacn:: ecase @induction_clause_list

   If the type of :n:`@term` has dependent premises, or dependent premises
   whose values are not inferrable from the :n:`with @bindings` clause,
   :n:`ecase` turns them into existential variables to be resolved later on.

.. tacn:: simple destruct {| @ident | @natural }

   Equivalent to :n:`intros until @ident; case @ident` when :n:`@ident`

>    is a quantified variable of the goal.

Author: @Zimmi48 do you know?  I sure don't.

>    is a quantified variable of the goal.

Reviewer: Testing gives:

```
No quantified hypothesis named m in current goal even after head-reduction.
```

.. tacv:: simple destruct @natural

   Behaves like :n:`intros until @natural; case @ident` where :n:`@ident`
   is the name given by :n:`intros until @natural` to the :n:`@natural` -th
   non-dependent premise of the goal.

.. tacn:: case_eq @one_term

   The tactic :n:`case_eq` is a variant of the :n:`case` tactic that allows
   performing case analysis on a term without completely forgetting its original
   form. This is done by generating equalities between the original form of the
   term and the outcomes of the case analysis.

.. tacn:: induction @induction_clause_list

   The argument :n:`@one_term` in :n:`@induction_clause_list` must be of

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: The induction hypothesis _always_ looks just like the theorem statement?  If so I maybe I should state that in a clear way.  Ex: `forall n, n + 0 = n` gives you the induction hypothesis `IHn : n + 0 = n`.

My initial thought was to look in the [Simple Inductive Types section](https://coq.inria.fr/distrib/current/refman/language/core/inductive.html#simple-inductive-types), where it talks about "elimination principles", which is barely described in the doc.  How does `nat_rec` relate to the induction hypothesis?

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer: >The induction hypothesis always looks just like the theorem statement?

No. It depends on the inductive type definition. `nat` is a very simple case.

>  How does `nat_rec` relate to the induction hypothesis?

It doesn't. `nat_ind` does.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer 2: Doesn't `induction` use `nat_rec` if your goal is in `Set` instead of `Prop`?

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer: Indeed, it does.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: I guess the most relevant thing would be to explain how the type of `nat_ind` from `Check nat_ind.` relates to the behavior of the tactic.  At least as an example.

```
nat_ind : 
  forall P : nat -> Prop,
  P 0 ->
  (forall n : nat, P n -> P (S n)) ->
  forall n : nat, P n
```

But it seems to me there be a lot that's unsaid about the various elimination principles that inductives generate, e.g. for `nat`:

```
nat is defined
nat_rect is defined
nat_ind is defined
nat_rec is defined
nat_sind is defined
```

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: Also, if `nat` is `Set`, then are the other 3 elimination principles ever (or commonly) useful for `nat`?

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer 2: The difference in induction principles is about the Sort of the goal you are proving, not about the Sort of `nat`.

`induction X using lem` is roughly the same as `generalize dependent X; apply lem; intros <whatever was generalized>`, modulo some checking about arguments to `lem`.  This should explain how the type of induction principles relates to the type of the goal and the types of the generated goals, roughly.  Suppose the goal is the type `G` and the inductive family of `X` is named `T` and that no `using` clause was passed; then `lem` defaults to `T_ind` if the Sort of `G` is `Prop`, `T_rec` when it is `Set`, `T_rect` if it is `Type`, and `T_sind` if the Sort of `G` is `SProp`.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: Also, I notice some of the doc uses the term "elimination principle", such as [here](https://coq.inria.fr/distrib/current/refman/language/core/inductive.html#simple-inductive-types).  Seems to me it would be more intuitive to refer to these as "induction principles" and add that term to the glossary.  WDYT?

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: Ah, the `using` clause in `induction` and `destruct` explicitly provides the induction principle?  If so, I will document that.  Pretty sure that wasn't in the original text.

Q1: `induction` and `destruct` are both needed? IIUC the difference between them is slight.  Which is used more/is more important?  Can we give any non-trivial guidance on when to use one or the other?  (IIUC, you must use `destruct` for co-induction.  See [here](https://coq.inria.fr/distrib/current/refman/language/core/coinductive.html#coq:cmd.CoInductive).)  Perhaps `induction` should appear in the doc before `destruct` (in part because `induction` is more obviously related to "Reasoning with inductive types")?  

Q2: Is it common to have a user-defined induction principle?  Is the purpose to do things like [prefix induction](https://en.wikipedia.org/wiki/Mathematical_induction#Prefix_induction)?

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer: >Also, I notice some of the doc uses the term "elimination principle", such as here. Seems to me it would be more intuitive to refer to these as "induction principles" and add that term to the glossary. WDYT?

I guess elimination principle is in theory more general (also applies when there is no recursion) but uniformizing toward induction principle could make sense in the context of Coq.

> Pretty sure that wasn't in the original text.

Here: https://coq.github.io/doc/master/refman/proofs/writing-proofs/reasoning-inductives.html#coq:tacv.induction-%E2%80%A6-using-%E2%80%A6

Q1: both are equally important. You don't want to generate induction hypotheses when you don't need them. You don't want your proof term to contain a `fix` when only a `match` was necessary. Just like you don't want to use `Fixpoint` when all you needed was `Definition`. I don't get your reference to co-induction (but co-induction is anyway a special case that is much less well-supported in Coq).

>  Perhaps `induction` should appear in the doc before `destruct` (in part because `induction` is more obviously related to "Reasoning with inductive types")?

No, `destruct` should appear first. In fact, we should probably separate the section in two "Case analysis" then "Induction". Here the title of the chapter is encompassing, in particular of non-recursive inductive types (for which `induction` is useless) and of co-inductive types (for which there is the `cofix` tactic).

Q2: Hugo gave an example in another thread for double induction. I guess the answer (of how common it is) is it depends on the users.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: I changed some of the uses of "elimination principle" to "induction principle" and made the latter a defined term, but I didn't change all 47 uses of the latter to use `:term:`--I expect the term is mostly well-understood.  Though I could certainly change them all.

That led me to wonder whether the "Generation of Induction Principles with Scheme" section might be better in the "Core Language/Inductive types" chapter, which already has quite a bit material about generation induction principles.  Also that section has a bunch of commands in it whereas the "Reasoning with inductive types" section seems like it should focus on tactics.  Also maybe move the "Generation of inversion principles" section to the other chapter.  What do you think?  Unfortunately, the Inductive types section is already fairly long.

Also, the explanation for the second example (odd and even) after [here](https://coq.inria.fr/distrib/current/refman/user-extensions/proof-schemes.html#coq:cmd.Scheme) seems incomplete--when would you use the induction principles `odd_ind`/`even_ind` versus `odd_even`/`even_odd`?  The latter would have to explicit on an `induction` tactic.  And the examples of induction principles should show their type more often and perhaps say just a little about the principle is generated (AFAICT one subterm per constructor).  Perhaps some of this could be changed outside of this PR.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: Hardly a model of clarity.  "term as induction scheme" doesn't make it clear that it refers to the second "term" in the grammar.  I'll change `induction scheme` to `induction principle` everywhere.

![image](https://user-images.githubusercontent.com/1253341/121289122-582fa100-c899-11eb-8922-5068dc9d64cf.png)

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer: > That led me to wonder whether the "Generation of Induction Principles with Scheme" section might be better in the "Core Language/Inductive types" chapter, which already has quite a bit material about generation induction principles. Also that section has a bunch of commands in it whereas the "Reasoning with inductive types" section seems like it should focus on tactics. Also maybe move the "Generation of inversion principles" section to the other chapter. What do you think? Unfortunately, the Inductive types section is already fairly long.

This would really go against the spirit of the "Core language" chapter which was meant to be focused on the kernel language, which the `Scheme` commands have nothing to do with (they have everything to do with reasoning with inductive types). I would rather go in the other direction and move part of the explanation currently in "Core language" to here (but we can think of that later).

> Also, the explanation for the second example (odd and even) after [here](https://coq.inria.fr/distrib/current/refman/user-extensions/proof-schemes.html#coq:cmd.Scheme) seems incomplete--when would you use the induction principles `odd_ind`/`even_ind` versus `odd_even`/`even_odd`?

Currently, auto-generated induction principles for mutual inductive types are useless, so they are never used. This could be spelled out more clearly in the doc.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: I added a statement about the useless IPs.

I notice that the tree/forest examples [here](https://coq.github.io/doc/master/refman/language/core/inductive.html#mutually-defined-inductive-types) and [here](https://coq.github.io/doc/master/refman/proofs/writing-proofs/reasoning-inductives.html#coq:cmd.Combined-Scheme) are almost identical.  I'll drop the first one but I'll use `Parameter`s rather than `Axiom`s.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: On second thought, I'll hold off on removing one of the near duplicates.

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Author: Would it make sense to enhance `Inductive` so it defines a useful induction principle for mutually inductive types?

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer 2: Yes.  I suspect most of the issue is just finding the time to write it, though it's possible there are some subtleties that need to be figured out

>    inductive type and the tactic :n:`induction` generates subgoals, one for

Reviewer: The main reason why it has been hold off was not a lack of time but the perfect-is-the-enemy-of-good problem. Coq developers wanted to figure out the most generic mutual induction principles before changing what `Inductive` generates. But this strategy has turned out to be highly detrimental to the users. So today, the developers would support anyone who would integrate the mutual induction principles that `Scheme` generates in the default mechanism used by `Inductive`.

   hypotheses are added to the local context using names whose prefix
   is **IH**.

>    There are several cases:

Author: Unless @Zimmi48 knows, I will use less specific language.

>    There are several cases:

Reviewer: I think it is. It makes sense to consider it is and fix it later if it turns out to be incomplete.

   + If :n:`@one_term` is an identifier :n:`@ident` denoting a quantified variable of the
     conclusion of the goal, then inductionident behaves like :n:`intros until
     @ident; induction @ident`. If :n:`@ident` is no longer dependent in the
     goal after application of :n:`induction`, it is erased (to avoid erasure,

>      use parentheses, as in :n:`induction (@ident)`).

Reviewer 2: I thought this only worked for hypotheses.  Does this trick really work for quantified variables of the goal too?  If so, induction is more magic than I thought it was.

.. exn:: Not an inductive product.
   :undocumented:

.. exn:: Unable to find an instance for the variables @ident … @ident.

   Use in this case the variant :tacn:`elim … with` below.

.. tacv:: induction @term @as_or_and_ipat

   Behaves like :tacn:`induction` but uses the names in
   :n:`@as_or_and_ipat` to name the variables introduced in the

>    context. The :n:`@as_or_and_ipat` must typically be of the form

Author: @Zimmi48, any comment?  @JasonGross?

>    context. The :n:`@as_or_and_ipat` must typically be of the form

Author: Or a simpler example:  Here I don't see how you would know that "m" and "IHX" are associated with `nat` and the hypothesis and not vice-versa, let alone for the 7 variable example above.

```
Goal forall n:nat, n + 0 = n.
induction n as [|m IHX]. 2: {
(* 1 subgoal
m : nat
IHX : m + 0 = m
______________________________________(1/1)
S m + 0 = S m *)

Check nat_ind.
(* nat_ind
     : forall P : nat -> Prop,
       P 0 ->
       (forall n : nat, P n -> P (S n)) ->
       forall n : nat, P n *)
```

>    context. The :n:`@as_or_and_ipat` must typically be of the form

Reviewer: One thing that is always easy to guess in the `as` clause structure for `destruct`, `induction` and `inversion` is the number of cases separated with `|` (the same as the number of constructors).
For `destruct`, it's also relatively easy to guess the number and the order of the variables because it corresponds to the number and order of arguments of each constructor.
For `induction`, this is a bit more complex, but you can still guess by looking at the form of the induction principle. E.g.:

```coq
nat_ind
     : forall P : nat -> Prop, P 0 -> (forall n : nat, P n -> P (S n)) -> forall n : nat, P n
```

There are two premises which correspond to the two constructors. The first premise has no (dependent or non-dependent) premise of its own, while the second has `n`, followed by `P n`, so you need to provide a name for the variable and a name for the (induction) hypothesis.

On the earlier more complex example:

```coq
clos_refl_trans_ind
     : forall (A : Type) (R : relation A) (P : A -> A -> Prop),
       (forall x y : A, R x y -> P x y) ->
       (forall x : A, P x x) ->
       (forall x y z : A, R* x y -> P x y -> R* y z -> P y z -> P x z) -> forall x a : A, R* x a -> P x a
```

The premise which corresponds to the third constructor is:

```coq
forall x y z : A, R* x y -> P x y -> R* y z -> P y z -> P x z
```

You have variables (dependent premises) followed by four hypotheses (non-dependent premises).

         induction y in x |-   *.
         Show 2.

.. tacv:: induction @term with @bindings @as_or_and_ipat using @term with @bindings in @goal_occurrences
          einduction @term with @bindings @as_or_and_ipat using @term with @bindings in @goal_occurrences

>    These are the most general forms of :tacn:`induction` and :tacn:`einduction`. It combines the

Reviewer 2: ```suggestion
   These are the most general forms of :tacn:`induction` and :tacn:`einduction`. They combine the
```

   effects of the with, as, using, and in clauses.

.. tacn:: elim @one_term_with_bindings {? using @one_term {? with @bindings } }

   This is a more basic induction tactic. Again, the type of the argument

>    :n:`@one_term` must be an inductive type. Then, according to the type of the

Reviewer: More basic / most similar to `induction`. The main difference between `elim` and `induction` and `case` and `destruct` is that `elim` and `case` do not introduce the variables in the goals they generate. There are other subtle differences but I don't know about them.

>    :n:`@one_term` must be an inductive type. Then, according to the type of the

Reviewer 2: Another difference is that `destruct` and `induction` generalize over hypotheses, whereas `case` (and I presume `elim` only look for occurrences in the goal.  Notably this means that if the term being case analyzed is a context variable, it's not cleared by `case`.

>    :n:`@one_term` must be an inductive type. Then, according to the type of the

Author: `fix` and `cofix` tactics are meant for use with `case` and/or `elim`.  They are a) not needed for `inductive` or `destruct` and b) not otherwise useful?

>    :n:`@one_term` must be an inductive type. Then, according to the type of the

Reviewer: I don't understand your question (if you are talking about the `fix` and `cofix` tactics or term constructs). In any case, `fix` is low level. You don't need to call it directly if you use `elim` or `induction` (which insert an induction principle which itself uses `fix` internally). `cofix` is different because there is not yet any high-level tactic which would allow doing without (in the stdlib, but people actually working with co-inductive types generally use external helper libraries).

   goal, the tactic ``elim`` chooses the appropriate destructor and applies it
   as the tactic :tacn:`apply` would do. For instance, if the local context
   contains :g:`n:nat` and the current goal is :g:`T` of type :g:`Prop`, then
   :n:`elim n` is equivalent to :n:`apply nat_ind with (n:=n)`. The tactic
   ``elim`` does not modify the context of the goal, neither introduces the
   induction loading into the context of hypotheses. More generally,
   :n:`elim @one_term` also works when the type of :n:`@one_term` is a statement
   with premises and whose conclusion is inductive. In that case the tactic
   performs induction on the conclusion of the type of :n:`@one_term` and leaves the
   non-dependent premises of the type as subgoals. In the case of dependent
   products, the tactic tries to find an instance for which the elimination

>    lemma applies and fails otherwise.

Author: Seems like more words than are necessary.  The first part seems to be that it does `apply nat_ind with (n:=n)` (a confusing choice of variables!).  But I don't quite get the part starting on new line 370 ("More generally ... performs induction ..."--the "performs induction" seems a bit circular!  Can you elaborate on this a little?

"premises" of A -> B -> C would be A and B?  Then what are the non-dependent premises here?

I want to get this clear because `simple induction` is defined in terms of `elim`.

Reviewer: > More generally ... performs induction ..."--the "performs induction" seems a bit circular!

Yes, IIUC, it's just defining this more general case in function of the less general case: "if you call me on something whose type is `A -> B -> C`, where `C` is an inductive type, I will introduce the hypotheses `A` and `B` and act on `C` as I described above".

`A` and `B` in this case are indeed the (non-dependent) premises. Dependent premises would be if there was a `forall`. Recall that `A -> B` is just `forall _ : A, B`. It's called non-dependent because the variable of type `A` does not appear in `B`.

Author: How does this look?

![image](https://user-images.githubusercontent.com/1253341/121597064-17e53580-c9f5-11eb-8229-d73c62cccaba.png)

And I suppose `forall x:nat, 1 = 0` is pointless.

Reviewer 2: > `A` and `B` are (non-dependent) premises of `A -> B -> C`

Is the premise the type, or the name?  That is, in `forall x : nat, x = x`, is the premise `x` or is it `nat`?

> `A` is a dependent premise of `A -> A \/ B`

I think this is wrong.  `A` is a dependent premise(?) of something like `forall A : Prop, A`, or `forall A : unit, A = A`.

Author: Well, the goal context shows types and `A -> B -> C` is a type, so the premise would be a type, too.  I think it's needlessly confusing to be utterly rigorous about names vs types significand.  In algebra, from "x = 2" you would say "x is a variable", "x is an integer" as well as "x is 2".  "x is a variable equal to 2" is unnatural unless you're introducing the concept of a variable to grade-school student.

> I think this is wrong.

Theo?  I'm just the scribe here.

Reviewer 2: > Theo? I'm just the scribe here.

I don't see anything Theo said that I interpret as claiming that `A` and `B` are dependent in `A -> B -> C`.  Is there some bit of text that he wrote that you're referring to here?

Author: I'm trying to guess how to describe things I don't know enough about--that is not at all easy to do.  It's probably better for Théo to respond directly first before making this a 3 way conversation.  But, nonetheless, what text would you give to replace everything I wrote above?

Author: How about this?

* `n` is a :term:`dependent premise` of `forall n:nat, n + 0 = n` because
  `n` appears both in the binder of the `forall` and in the quantified statement
  `n + 0 = n`.
  
* `A` and `B` are (non-dependent) :gdef:`premises <premise>` of `A -> B -> C`.
  (Recall that `A -> B` is notation for the term `forall _ : A, B`; the wildcard
  `_` can't be used in the quantified statement.)

Reviewer: Indeed, Jason is right here and your last proposal looks better (and more correct) than the former.

Indeed, there is an ambiguity regarding what premise refers to (the term or its type). I think in the case of non-dependent premises, it really means the type, because this gives similar language to logic without dependent types. In the case of dependent premises, maybe we can write "`n : nat` is a dependent premise"?

.. tacv:: elim @one_term with @bindings
   :name: elim … with

>    Allows to give explicit instances to the premises of the type of :n:`@one_term`

Reviewer 2: ```suggestion
   Allows giving explicit instances to the premises of the type of :n:`@one_term`
```

.. seealso:: :tacn:`functional induction`

.. tacn:: discriminate {? @destruction_arg }

   This tactic proves any goal from an assumption stating that two
   structurally different :n:`@term`\s of an inductive set are equal. For
   example, from :g:`(S (S O))=(S O)` we can derive by absurdity any

>    proposition.

Author: The behavior is quirky.  I expected `0 = 1` or `1 <> 1` would derive a contradiction whether in a hypothesis or the goal.  But I get this:

```
Goal 1 = 2 -> True.
discriminate.   (* No more goals *)

Goal True -> 1 = 2.
discriminate.   (* No primitive equality found.  I expected to see 1 = 2 replaced with False *)

Goal 1 <> 1 -> True.
discriminate.   (* No primitive equality found. *)

Goal True -> 1 <> 2.
discriminate.   (* No more goals -- but 1 <> 2 is a true statement, didn't expect it to work on this one *)
```

Seems to me this a usability issue, perhaps worth fixing.  It may break existing proofs, though.

Reviewer 2: Maybe what you're missing is that `x <> y` is the same as `x = y -> False`.  The behavior is exactly what the doc states: if, after introducing all hypotheses, there is a contradictory equality in the context, then `discriminate` will prove any goal from it.  The tactic will only make progress if it solves the goal, so it's not going to replace any of your goals with `False` (this is not really useful behavior anyway).  Additionally, `discriminate` is about discriminating different things, not about instantiating hypotheses which posit that two equal things are not equal; it's not about instantiating hypotheses with reflexive proofs.

Perhaps you're thinking of `congruence`, which is a more general equality reasoning tactic?

Author: Yeah, I forgot about what's behind some of the basic Prop notations.  There should be a short section somewhere in the doc that explains those details (e.g. True, False, ->, <>, the things in Logic.v) because they seem important for users to know about.  Any idea where to put something like that?

No, not thinking about `congruence`.

Reviewer: There's https://coq.inria.fr/refman/language/coq-library.html

   This does the same thing as :n:`discriminate @term` but using the given
   bindings to instantiate parameters or hypotheses of :n:`@term`.

.. tacn:: ediscriminate {? @destruction_arg }
          ediscriminate {? @destruction_arg }
   :name: ediscriminate; _

   This works the same as :tacn:`discriminate` but if the type of :token:`one_term`, or the
   type of the hypothesis referred to by :token:`natural`, has uninstantiated
   parameters, these parameters are left as existential variables.

.. tacv:: discriminate

   This behaves like :n:`discriminate @ident` if ident is the name of an
   hypothesis to which ``discriminate`` is applicable; if the current goal is of
   the form :n:`@term <> @term`, this behaves like
   :n:`intro @ident; discriminate @ident`.

   .. exn:: No discriminable equalities.
      :undocumented:

.. tacn:: injection {? @destruction_arg } {? as {* @simple_intropattern } }

   The injection tactic exploits the property that constructors of
   inductive types are injective, i.e. that if :n:`c` is a constructor of an
   inductive type and :n:`c t__1` and :n:`c t__2` are equal then

>    :n:`t__1` and :n:`t__2` are equal too.

Author: I'm surprised this gives an error.  I expected it would simplify the goal to 0 = 1 or somehow indicate that there's a contradiction.  Is it a bug or was the description wrong/incomplete?

```
Goal True -> 1 = 2.
intros.
injection.  (* Not a negated primitive equality. *)
```

The example in the original text [here](https://coq.inria.fr/distrib/current/refman/proof-engine/tactics.html#coq:tacn.injection) doesn't match the text very well.  The found equalities are stacked up as `E1 -> E2 -> ...`

```
  H0 : cons n l = cons 0 nil
  ============================
  P l

injection H0.

  ============================
  l = nil -> n = 0 -> P l

Reviewer 2: >  Is it a bug or was the description wrong/incomplete?

The description neglects to mention that injection only works on hypotheses (whether in the context or quantified in the goal), not the conclusion.

      behavior for objects that are proofs of a statement in :g:`Prop`. This flag
      controls this behavior.

.. tacn:: simple injection {? @destruction_arg }
   :undocumented:

> .. tacn:: inversion {| @ident | @natural } using @one_term {? in {+ @ident } }

Reviewer: This most likely is a limitation that is not justified (by anything else than "not implemented").

> .. tacn:: inversion {| @ident | @natural } using @one_term {? in {+ @ident } }

Reviewer 2: The `using` clause makes no sense for `inversion_sigma` (which is basically just picking a smarter induction principle to pass to the `using` clause of `induction` for equalities of sigmas).  I believe @Zimmi48 is right about the others though.

          inversion {| @ident | @natural } {? @as_or_and_ipat } {? in {+ @ident } }
   :name: inversion; _

   .. insertprodn as_or_and_ipat as_or_and_ipat

   .. prodn::
      as_or_and_ipat ::= as {| @or_and_intropattern | @ident }
      | as @equality_intropattern

   Let the type of :n:`@ident` in the local context be :g:`(I t)`, where :g:`I`

>    is a (co)inductive predicate. Then, ``inversion`` applied to :n:`@ident`

Reviewer: Inversion is one of the hardest tactic to grasp for my students. It allows performing a case analysis when one of the hypothesis in the context has an inductive type. The case analysis will show all the possible ways we could obtain this hypothesis using the constructors of the inductive type and taking into account the shape of the arguments it is applied to. I can provide examples if that's helpful.

>    is a (co)inductive predicate. Then, ``inversion`` applied to :n:`@ident`

Reviewer 2: I think one of the things that makes `inversion` subtle is that it's behavior seems very different on single-conductor inductive families (like equalities such as `(x, y) = (1, 2)`) and on multi-constructor inductives without indices such as `list nat`.  The commonality only becomes apparent when using multi-constructor inductives with indices, which, IIRC, has `Vector.t` as the only example in the standard library.  (I believe ASTs are another common example, as are dependent well-formedness conditions.)

>    is a (co)inductive predicate. Then, ``inversion`` applied to :n:`@ident`

Author: I'd like to get a much better description of inversion with examples, but I think that will have to wait for a future PR.

   where :n:`@ident` is the identifier for the last introduced hypothesis.

.. tacv:: inversion_clear @ident

   Behaves like :n:`inversion` and then erases :n:`@ident` from the context.

.. tacv:: inversion @ident @as_or_and_ipat

   This generally behaves like inversion but using names in :n:`@as_or_and_ipat`
   for naming hypotheses. The :n:`@as_or_and_ipat` must have the form

>    :n:`[p__11 … p__1n | … | p__m1 … p__mn]`

Author: This paragraph is long and dense.  I think some of it is repeating the explanation of how `@as_or_and_ipat` works, which should be explain in one place (not here).  Perhaps you can help me figure out what parts are unique to this tactic?

Reviewer: Most of it is specific to the `inversion` tactic. And honestly, I don't think that many users are actually using this form because the variables that `inversion` generate are quite unpredictable.

   with `m` being the number of constructors of the type of :n:`@ident`. Be
   careful that the list must be of length `m` even if ``inversion`` discards
   some cases (which is precisely one of its roles): for the discarded

>    cases, just use an empty list (i.e. `n = 0`).The arguments of the i-th

Reviewer 2: ```suggestion
   cases, just use an empty list (i.e. `n = 0`). The arguments of the i-th
```

   constructor and the equalities that ``inversion`` introduces in the
   context of the goal corresponding to the i-th constructor, if it

>    exists, get their names from the list :n:`p__i1 … p__in` in

Reviewer 2: This sentence is just part of how `@as_or_and_ipat` works

>    order. If there are not enough names, ``inversion`` invents names for the

Reviewer 2: This sentence is also just part of how `@as_or_and_ipat` works.

   remaining variables to introduce. In case an equation splits into several
   equations (because ``inversion`` applies ``injection`` on the equalities it
   generates), the corresponding name :n:`p__ij` in the list must be
   replaced by a sublist of the form :n:`[p__ij1 … p__ijk]`
   (or, equivalently, :n:`(p__ij1 , …, p__ijk)`) where
   `k` is the number of subequalities obtained from splitting the original
   equation. Here is an example. The ``inversion … as`` variant of

>    ``inversion`` generally behaves in a slightly more expectable way than

Reviewer 2: Use "expected" or "intuitive"

   ``inversion`` (no artificial duplication of some hypotheses referring to

>    other hypotheses). To take benefit of these improvements, it is enough to use

Reviewer 2: "take benefit" is not standard English

>    ``inversion … as []``, letting the names being finally chosen by Coq.

Reviewer 2: ```suggestion
   ``inversion … as []``, letting the names be finally chosen by Coq.
```

   Let :n:`{+ @ident}` be identifiers in the local context. This tactic behaves
   as generalizing :n:`{+ @ident}`, and then performing ``inversion_clear``.

.. tacv:: inversion_clear @ident @as_or_and_ipat in {+ @ident}

   This allows naming the hypotheses introduced in the context by
   :n:`inversion_clear @ident in {+ @ident}`.

.. todo PR: expand "dependent" tactic in editing

> .. tacn:: dependent {| simple inversion | inversion | inversion_clear } {| @ident | @natural } {? @as_or_and_ipat } {? with @one_term }

Author: Q1: I suppose the options `simple inversion`, `inversion` and `inversion_clear` should be presented separately?  (The mlg file has them combined like this).

Q2: There are many tactics that contain "dependent" and "simple" (listed below).  Is there any commonality between the various "dependent" tactics or between the "simple" tactics, or do they mean different things in different tactics?

```
simple destruct {| @ident | @natural }
simple induction {| @ident | @natural }
simple injection {? @destruction_arg }
simple inversion {| @ident | @natural } {? @as_or_and_ipat } {? in {+ @ident } }
dependent induction @ident {? {| generalizing | in } {+ @ident } } {? using @one_term }
dependent {| simple inversion | inversion | inversion_clear }
     {| @ident | @natural } {? @as_or_and_ipat } {? with @one_term }
```

Reviewer: Q1: yes, I guess this would be more readable if they could be separated.

Q2: yes, I think there are some commonalities, although I couldn't explain precisely, because I do not use any of these `simple` / `dependent` variants.

Reviewer 2: No idea on Q1, but for Q2, I think (though I'm not sure) that "simple" and "dependent" share roughly the same meanings across these tactics (though `simple refine`, for example, is very different).

.. tacv:: simplify_eq

   If the current goal has form :g:`t1 <> t2`, it behaves like
   :n:`intro @ident; simplify_eq @ident`.

.. tacn:: dependent rewrite {? {| -> | <- } } @one_term {? in @ident }

>    :name: dependent rewrite ->

Author: Should this be in the "Reasoning with equalities" chapter?

Reviewer: Yes, that could make sense. But a pointer to this tactic should appear in this chapter given the note that this kind of equalities may be derived by the `injection` and `inversion` tactics.

Reviewer: Ah, on the other hand, this is already part of the "Equality and inductive sets" section which is entirely on the fringe between the two chapters.

Author: Your final answer is "keep it in this chapter", but (I expect) add a hyperlink to it in the equality chapter?

Reviewer: No, I don't have a final answer. But if you move it, then add a link.

>    .. todo PR: what is "existT" below?

Reviewer 2: The constructor of `sigT`, defined in the standard library

Author: OK, thanks.  Still mysterious, but I'll let it go for now.

Off topic: I've noticed that it's difficult to find the definitions of things like `sigT` and `existT` with a global text search because people reuse the identifier.  How useful would it be if you could shift-click on any identifier or notation symbol in the GUI to jump to its definition?  Or to search for all uses of a definition in a given file or perhaps globally?

Reviewer 2: https://github.com/coq/coq/blob/e60305c12305d298fe7e36738b8c95eca3d7409e/theories/Init/Specif.v#L43-L45

> How useful would it be if you could shift-click on any identifier or notation symbol in the GUI to jump to its definition? Or to search for all uses of a definition in a given file or perhaps globally?

Those both sound useful, though I personally primarily use ProofGeneral.  I think another useful thing would be links between the refman and the stdlib.

Author: For jumping to the definition, you'd do most of the work on the Coq side so it can be shared among multiple GUIs. So should be easy for PG to support it.
